src/lpython/semantics/python_comptime_eval.h

```diff
@@ -163,6 +166,64 @@ struct PythonIntrinsicProcedures {
                 ASRUtils::type_to_str_python(arg_type) + "'", loc);
         }
     }
+    static ASR::expr_t *eval_upper(Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args) {
+        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
+        if (args.size() != 1) {
+            throw SemanticError("upper() takes exactly one argument (" +
+                std::to_string(args.size()) + " given)", loc);
+        }
+        ASR::ttype_t* str_type = ASRUtils::TYPE(ASR::make_Character_t(al,loc, 1, 1, nullptr, nullptr, 0));
```

Comment:
I think the return type should be same as input type which means you can just do,

```suggestion
        ASR::ttype_t* str_type = ASRUtils::expr_type(args[0]);
```